diff -U 2 -H -b -w -B -E -d -p -r -N -x '*.svn' -x '*.d' -- svn_hcraft_1.2/irrlicht/include/IGUIEditBox.h hcraft_libs/irrlicht/include/IGUIEditBox.h
--- svn_hcraft_1.2/irrlicht/include/IGUIEditBox.h	2006-06-27 12:46:42.000000000 +0100
+++ hcraft_libs/irrlicht/include/IGUIEditBox.h	2007-02-08 13:05:57.000000000 +0100
@@ -55,4 +55,18 @@ namespace gui
 		//! Returns maximum amount of characters, previously set by setMax();
 		virtual s32 getMax() = 0;
+
+        // MICHA: useful addition
+		virtual void setUnusableChars(const wchar_t* text)
+		{
+		    UnusableChars = text;
+		}
+
+		const wchar_t* getUnusableChars(const wchar_t* text)
+		{
+		    return UnusableChars.c_str();
+        }
+
+    protected:
+        core::stringw UnusableChars;
 	};
 
diff -U 2 -H -b -w -B -E -d -p -r -N -x '*.svn' -x '*.d' -- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIEditBox.cpp hcraft_libs/irrlicht/source/Irrlicht/CGUIEditBox.cpp
--- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIEditBox.cpp	2006-12-04 16:19:49.000000000 +0100
+++ hcraft_libs/irrlicht/source/Irrlicht/CGUIEditBox.cpp	2007-05-31 02:11:55.000000000 +0100
@@ -22,5 +22,5 @@ CGUIEditBox::CGUIEditBox(const wchar_t* 
 						 IOSOperator* op)
 : IGUIEditBox(environment, parent, id, rectangle), Border(border), 
-	OverrideColorEnabled(false), OverrideFont(0), CursorPos(0), ScrollPos(0),
+    OverrideColorEnabled(false), OverrideFont(0), BlinkStartTime(0), CursorPos(0), ScrollPos(0),
 	MarkBegin(0), MarkEnd(0), MouseMarking(false), Operator(op), Max(0)
 {
@@ -88,5 +88,5 @@ bool CGUIEditBox::OnEvent(SEvent event)
 			MarkBegin = 0;
 			MarkEnd = 0;
-			return true;
+			break;
 		}
 		break;
@@ -209,4 +209,8 @@ bool CGUIEditBox::processKey(const SEven
 	switch(event.KeyInput.Key)
 	{
+    case KEY_UP:
+    case KEY_DOWN:
+    case KEY_TAB:
+        return false;
 	case KEY_END:
 		if (event.KeyInput.Shift)
@@ -370,4 +374,7 @@ bool CGUIEditBox::processKey(const SEven
 		if (event.KeyInput.Char != 0)
 		{
+		    if ( IsCharUnusable(event.KeyInput.Char) )
+                return false;
+
 			if (Text.size() < Max || Max == 0)
 			{
@@ -446,4 +453,14 @@ bool CGUIEditBox::processKey(const SEven
 }
 
+bool CGUIEditBox::IsCharUnusable(wchar_t c)
+{
+    for ( int i=0; i < UnusableChars.size(); ++i )
+    {
+        if ( c == UnusableChars[i] )
+            return true;
+    }
+    return false;
+}
+
 //! draws the element and its children
 void CGUIEditBox::draw()
diff -U 2 -H -b -w -B -E -d -p -r -N -x '*.svn' -x '*.d' -- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIEditBox.h hcraft_libs/irrlicht/source/Irrlicht/CGUIEditBox.h
--- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIEditBox.h	2006-12-04 16:19:49.000000000 +0100
+++ hcraft_libs/irrlicht/source/Irrlicht/CGUIEditBox.h	2007-02-08 13:10:50.000000000 +0100
@@ -55,4 +55,7 @@ namespace gui
 	protected:
 
+        // MICHA: useful addition
+        bool IsCharUnusable(wchar_t c);
+
 		bool processKey(const SEvent& event);
 		bool processMouse(const SEvent& event);
