diff -U 2 -H -b -w -B -E -d -p -r -N -x '*.svn' -x '*.d' -- svn_hcraft_1.2/irrlicht/include/IGUIStaticText.h hcraft_libs/irrlicht/include/IGUIStaticText.h
--- svn_hcraft_1.2/irrlicht/include/IGUIStaticText.h	2006-06-27 12:46:42.000000000 +0100
+++ hcraft_libs/irrlicht/include/IGUIStaticText.h	2006-07-06 20:57:32.000000000 +0100
@@ -58,4 +58,7 @@ namespace gui
 		 \return Returns height of text in pixels. */
 		virtual s32 getTextHeight() = 0;
+
+        //! center text horicontally if no wordwrap is used
+		virtual void setHoricontalCenter(bool enable) = 0;
 	};
 
diff -U 2 -H -b -w -B -E -d -p -r -N -x '*.svn' -x '*.d' -- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIStaticText.cpp hcraft_libs/irrlicht/source/Irrlicht/CGUIStaticText.cpp
--- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIStaticText.cpp	2006-12-04 16:19:49.000000000 +0100
+++ hcraft_libs/irrlicht/source/Irrlicht/CGUIStaticText.cpp	2006-09-29 17:44:55.000000000 +0100
@@ -23,5 +23,5 @@ CGUIStaticText::CGUIStaticText(const wch
 : IGUIStaticText(environment, parent, id, rectangle), Border(border),
 	OverrideFont(0), OverrideColorEnabled(false), WordWrap(false),
-	LastBreakFont(0), Background(background)
+	LastBreakFont(0), Background(background), HoricontalCenter(false)
 {
 	#ifdef _DEBUG
@@ -81,5 +81,5 @@ void CGUIStaticText::draw()
 				font->draw(Text.c_str(), frameRect, 
 					OverrideColorEnabled ? OverrideColor : skin->getColor(IsEnabled ? EGDC_BUTTON_TEXT : EGDC_GRAY_TEXT),
-					false, true, &AbsoluteClippingRect);
+					HoricontalCenter, true, &AbsoluteClippingRect);
 			else
 			{
@@ -94,5 +94,5 @@ void CGUIStaticText::draw()
 					font->draw(BrokenText[i].c_str(), r,
 						OverrideColorEnabled ? OverrideColor : skin->getColor(IsEnabled ? EGDC_BUTTON_TEXT : EGDC_GRAY_TEXT),
-						false, false, &AbsoluteClippingRect);
+						HoricontalCenter, false, &AbsoluteClippingRect);
 
 					r.LowerRightCorner.Y += height;
@@ -274,5 +274,8 @@ s32 CGUIStaticText::getTextHeight()
 }
 
-
+void CGUIStaticText::setHoricontalCenter(bool enable)
+{
+    HoricontalCenter = enable;
+}
 
 } // end namespace gui
diff -U 2 -H -b -w -B -E -d -p -r -N -x '*.svn' -x '*.d' -- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIStaticText.h hcraft_libs/irrlicht/source/Irrlicht/CGUIStaticText.h
--- svn_hcraft_1.2/irrlicht/source/Irrlicht/CGUIStaticText.h	2006-06-27 12:46:39.000000000 +0100
+++ hcraft_libs/irrlicht/source/Irrlicht/CGUIStaticText.h	2006-07-06 20:58:04.000000000 +0100
@@ -48,4 +48,7 @@ namespace gui
 		virtual s32 getTextHeight();
 
+		//! center text horicontally if no wordwrap is used
+		virtual void setHoricontalCenter(bool enable);
+
 	private:
 
@@ -57,4 +60,5 @@ namespace gui
 		bool WordWrap;
 		bool Background;
+		bool HoricontalCenter;
 
 		video::SColor OverrideColor;
